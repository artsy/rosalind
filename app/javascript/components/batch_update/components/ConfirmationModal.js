import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { colors } from './Layout'
import { ESC } from 'lib/keycodes.js'
import { Button } from '@artsy/palette'

const Modal = styled.div`
  position: fixed;
  left: 50%;
  top: 40%;
  width: 600px;
  height: auto;
  background: white;
  z-index: 2;
  padding: 20px;

  // modal is closed
  visibility: hidden;
  opacity: 0;
  transform: scale(0.95) translate(-50%, -50%);
  transition: all 0.25s;

  // modal is open
  &.modal-open {
    visibility: visible;
    opacity: 1;
    transform: scale(1) translate(-50%, -50%);
    transition: all 0.125s;
  }

  & h1 {
    font-weight: normal;
    font-size: 1.5em;
    color: ${colors.grayDarker};
  }

  & section {
    color: ${colors.grayDarkest};
    border-top: solid 1px ${colors.grayLighter};
    margin-top: 1em;
    padding-top: 1em;
  }

  & section p {
    margin-bottom: 1em;
  }
`
Modal.displayName = 'Modal'

const Controls = styled.div`
  border-top: solid 1px ${colors.grayLighter};
  margin-top: 1em;
  padding-top: 1em;
  display: flex;
  flex-direction: row;
  justify-content: flex-end;

  & a {
    margin-left: 1em;
  }
`
Controls.displayName = 'Controls'

class ConfirmationModal extends React.Component {
  constructor(props) {
    super(props)
    this.handleKeyup = this.handleKeyup.bind(this)
    this.handleDismissClick = this.handleDismissClick.bind(this)
    this.handleAcceptClick = this.handleAcceptClick.bind(this)
  }

  componentWillUpdate({ isOpen: willBeOpen }) {
    const { isOpen } = this.props
    if (!isOpen && willBeOpen) {
      // modal is opening
      window.addEventListener('keyup', this.handleKeyup)
    }
    if (isOpen && !willBeOpen) {
      // modal is closing
      window.removeEventListener('keyup', this.handleKeyup)
    }
  }

  handleKeyup(e) {
    if (e.keyCode === ESC) {
      this.props.onDismiss()
    }
  }

  handleDismissClick(e) {
    e.preventDefault()
    this.props.onDismiss()
  }

  handleAcceptClick(e) {
    e.preventDefault()
    this.props.onAccept()
  }

  render() {
    const { isOpen, children } = this.props
    const className = isOpen ? 'modal-open' : null
    return (
      <Modal className={className}>
        {children}
        <Controls>
          <Button
            variant="secondaryOutline"
            mx={1}
            className="dismiss"
            onClick={this.handleDismissClick}
          >
            Go back
          </Button>
          <Button className="accept" onClick={this.handleAcceptClick}>
            Continue
          </Button>
        </Controls>
      </Modal>
    )
  }
}

ConfirmationModal.propTypes = {
  isOpen: PropTypes.bool,
  onDismiss: PropTypes.func.isRequired,
  onAccept: PropTypes.func.isRequired,
}

export default ConfirmationModal
