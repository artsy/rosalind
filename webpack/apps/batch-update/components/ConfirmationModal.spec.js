import React from 'react'
import renderer from 'react-test-renderer'
import ConfirmationModal from './ConfirmationModal'
import { mount } from 'enzyme'

let props

beforeEach(() => {
  props = {
    onDismiss: jest.fn(),
    onAccept: jest.fn()
  }
})

it('renders correctly when closed', () => {
  const rendered = renderer.create(
    <ConfirmationModal {...props}>
      I am closed
    </ConfirmationModal>
  )
  const tree = rendered.toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly when open', () => {
  const rendered = renderer.create(
    <ConfirmationModal {...props} isOpen>
      I am open
    </ConfirmationModal>
  )
  const tree = rendered.toJSON()
  expect(tree).toMatchSnapshot()
})

describe('when the "Go back" button is clicked', () => {
  it('calls the correct handler', () => {
    const wrapper = mount(
      <ConfirmationModal {...props}>
        Dismiss me?
      </ConfirmationModal>
    )
    const mockClickEvent = { preventDefault: jest.fn() }
    wrapper.find('a.dismiss').simulate('click', mockClickEvent)
    expect(props.onDismiss.mock.calls.length).toEqual(1)
  })
})

describe('when the "Continue" button is clicked', () => {
  it('calls the correct handler', () => {
    const wrapper = mount(
      <ConfirmationModal {...props}>
        Accept me?
      </ConfirmationModal>
    )
    const mockClickEvent = { preventDefault: jest.fn() }
    wrapper.find('a.accept').simulate('click', mockClickEvent)
    expect(props.onAccept.mock.calls.length).toEqual(1)
  })
})
