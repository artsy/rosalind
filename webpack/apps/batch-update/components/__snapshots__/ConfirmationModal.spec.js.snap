exports[`test renders correctly when closed 1`] = `
<div
  className="gungoz">
  I am closed
  <div
    className="kriMBh">
    <a
      className="dismiss jRkNFJ"
      href="#"
      onClick={[Function]}>
      Go back
    </a>
    <a
      className="accept jhxkip"
      href="#"
      onClick={[Function]}>
      Continue
    </a>
  </div>
</div>
`;

exports[`test renders correctly when open 1`] = `
<div
  className="modal-open gungoz">
  I am open
  <div
    className="kriMBh">
    <a
      className="dismiss jRkNFJ"
      href="#"
      onClick={[Function]}>
      Go back
    </a>
    <a
      className="accept jhxkip"
      href="#"
      onClick={[Function]}>
      Continue
    </a>
  </div>
</div>
`;
